[ADD] Estate Module (Python Framework Training)

OmarKhalil2001 opened the pull request:
[ADD] Implementation for Estate Module of Python Framework

Description:

This Pull Request introduces the Estate Module, which houses the new Estate FelBeit application. This application is designed to provide real estate agents with a centralized system for managing property listings, tracking the sales lifecycle, and handling the relationship between sellers and buyers.

Key Features:
1. Seller Tracking: Associates each property with a seller for streamlined management.
2. Buyer Tracking: Records potential buyers and formally links the successful buyer to the property upon closing.
3. Property Lifecycle Tracking: Provides a state-management system to monitor property stages from initial listing through offer negotiations to the final sale.
4. Offer Management: Enables agents to record offers, track their status, and perform actions to either accept or refuse them.
5. Spatial Calculations: Includes a precision tool to calculate indoor and outdoor areas separately, ensuring accurate property data.

OmarKhalil2001 committed:
[ADD] Chapter 2: Adding Real Estate Module. Defining Manifest and Init

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1068.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1068)

OmarKhalil2001 committed:
[IMP] Adding Buildings Model to Estate Module

OmarKhalil2001 committed:
[IMP] Adding an access control for estate module

barracudapps commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,11 @@
+{
+    'name': 'real estate',
+    'category': 'Tutorials',
+    'version': '1.0',
```
Maybe adding a small description here might be interesting

barracudapps commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,11 @@
+    'category': 'Tutorials',
+    'version': '1.0',
+    'application': True,
+    'installable': True,
```
Normally `installable` is true by default

barracudapps commented on estate/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import models
```
Don't forget Unix conventions on text files' structure based on a line definition (cf. https://peps.python.org/pep-0008/ and https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206)
```suggestion
from . import models

```

Please check all your files.

barracudapps commented on estate/models/buildings_model.py:
```diff
@@ -0,0 +1,17 @@
+
+class buidlings_model(models.Model):
+    _name = 'estate.buildings'
+    _description = 'Buildings Model'
```
Please use double quotes for non-technical strings
```suggestion
    _description = "Buildings Model"
```

barracudapps commented on estate/models/buildings_model.py:
```diff
@@ -0,0 +1,17 @@
+from odoo import models, fields   # noqa: F401
+
+class buidlings_model(models.Model):
```
Rename your model accordingly based on the module
Some interesting docs:

- https://www.odoo.com/documentation/18.0/contributing/development/coding_guidelines.html?highlight=naming#symbols-and-conventions
- https://peps.python.org/pep-0008/#naming-conventions

barracudapps commented on estate/models/buildings_model.py:
```diff
@@ -0,0 +1,17 @@
+        ('south', 'South'),
+        ('east', 'East'),
+        ('west', 'West')
+    ], string='Garden Orientation')
```
Functionally, it would make more sense to have the string parameter as first argument. This will also allow you to remove its naming by simply writing the string itself like this:
```suggestion
    garden_orientation = fields.Selection(
        "Garden Orientation",
        selection=[
            ('north', 'North'),
            ('south', 'South'),
            ('east', 'East'),
            ('west', 'West'),
        ],
    )
```

Refer to the `Field` class for more information (https://github.com/odoo/odoo/blob/0a97fa64783e4b58bb17db79343fb566196e71eb/odoo/orm/fields.py#L92)

barracudapps reviewed (changes requested):
Some comments here and there. Can you also please review your PR title and description following our guidelines (https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html)?
A clear commit message is really important to describe what you did and why you did it.

OmarKhalil2001 committed:
[IMP] Completed Chapter 5: Added Basic GUI to Estate Module

OmarKhalil2001 committed:
[IMP] Chapter 6: Addin Filters, Group By, Search, and Forms.

OmarKhalil2001 committed:
[IMP] Chapter 7: Added Tags, Buyer, Seller, and Offer models to Estate Module. Improved UI to show all the details about  each building

barracudapps commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,10 @@
+    "category": "Tutorials",
+    "version": "1.1",
+    "application": True,
+    "data": ["security/ir.model.access.csv", "views/views.xml", "views/menus.xml"],
```
Please split your data files (same comment for as many lists as possible):
```suggestion
    "data": [
        "security/ir.model.access.csv",
        "views/views.xml",
        "views/menus.xml",
    ],
```

If someone has to add a file, it will be way easier and keep a clear overview (with git history) of the changes.

barracudapps commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,10 @@
+    "version": "1.1",
+    "application": True,
+    "data": ["security/ir.model.access.csv", "views/views.xml", "views/menus.xml"],
+    "author": "OMKHA",
```
```suggestion
    "author": "Odoo",
```
As you are developing for Odoo, the company is the author of the file.

barracudapps commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,10 @@
+{
+    "name": "real estate",
+    "description": "real estate renting management system",
```
I would have capitalized these lines

barracudapps commented on estate/models/building_type_model.py:
```diff
@@ -0,0 +1,8 @@
+from odoo import models, fields
+
+
+class building_type_model(models.Model):
```
The class name should match the file name (for most cases) and follow the PascalCase convention.
As PEP8 says (https://peps.python.org/pep-0008/#class-names): 
> Class names should normally use the CapWords convention

```suggestion
class BuildingType(models.Model):
```

barracudapps commented on estate/models/buildings_model.py:
```diff
@@ -0,0 +1,38 @@
+    description = fields.Text()
+    value = fields.Integer(copy=False)
+    availability_date = fields.Date(
+        default=fields.Date.today() + timedelta(days=90), copy=False
```
Use a lambda function or call a compute method because this will fix the default date to the server starting date

barracudapps commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+access_first_model,access_first_model,model_estate_buildings,base.group_user,1,1,1,1
+access_building_type_model,access_building_type_model,model_estate_building_type,base.group_user,1,1,1,1
+access_building_tags_model,access_building_tags_model,model_estate_building_tags,base.group_user,1,1,1,1
+access_offers_model,access_offers_model,model_estate_offers,base.group_user,1,1,1,1
```
Don't forget Unix conventions on text files' structure based on a line definition (cf. https://peps.python.org/pep-0008/ and https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206)

Check all XML files too

barracudapps commented on estate/views/views.xml:
```diff
@@ -0,0 +1,85 @@
+                <field name="name"/>
+                <field name="value"/>
+                <separator/>
+                <filter string="Available" name="Available" domain="['|', ('state', '=', 'new' ), ('state', '=', 'offer received')]"/>
```
Maybe try something like `[('state', 'in', ['new', 'offer_received'])]`

barracudapps reviewed (changes requested):
Thanks for the feature. Here are some comments.

Avoid as much redundancy as possible, ease the searching and respect our naming conventions.
Let's take `buildings_model.py` as an example. The file is in the `models` folder of the `estate` module a better naming would have been `building.py` (or maybe `estate_building.py` as it allow a quick search on the module name and `building.py` might exist in many modules with different meanings).
Moreover, as it is a model the name should not be plural. Indeed, this is the template to create one building (multiple times but it's still 1 item).
Last thing here: please follow the instructions as if the task (tutorial) was written by a PO). You can of course challenge everything but I would've gone with a `property` instead of a `building`.

Also, can you please change the PR message and tell a bit more about what the added code is doing and why you added these lines?

OmarKhalil2001 committed:
[IMP]  Chapter 8: Added  validity, date_deadline attributes for offers model. improved UX with compute and inverse methods

OmarKhalil2001 committed:
[IMP] Added Accept and Refuse buttons to offers. Added Sell and Cancel buttons to Buildings Model in Estate Module.

OmarKhalil2001 committed:
[CLN] Addressing Clean Code comments for Estate Module. Added newlines to the end of files. Improved the readability of manifest. renamed model classes to follow the pascal casing.

barracudapps commented on estate/models/building.py:
```diff
@@ -0,0 +1,83 @@
+
+
+class Building(models.Model):
+    _name = "estate.buildings"
```
Good job on your refactoring but still some small consistency issues like here (and the file names)
```suggestion
    _name = 'estate.building'
```

barracudapps commented on estate/models/buildings_model.py:
```diff
@@ -0,0 +1,83 @@
+    state = fields.Selection(
+        [
+            ("new", "New"),
+            ("offer received", "Offer Received"),
```
Use snake_case for variables and technical keys
```suggestion
            ('offer_received', "Offer Received"),
```

barracudapps commented on estate/models/buildings_model.py:
```diff
@@ -0,0 +1,83 @@
+    def action_set_sold(self):
+        for record in self:
+            if record.state == "canceled":
+                raise UserError("Canceled buildings cannot be sold.")
```
Please ensure displayed texts can be translated. You can use the `self.env._` method to declare translatable terms

barracudapps commented on estate/models/offers_model.py:
```diff
@@ -0,0 +1,63 @@
+
+class Offer(models.Model):
+    _name = "estate.offers"
+    _description = "Offers Model"
```
Up to you but I would avoid redundancy by removing `Model` here

barracudapps reviewed (changes requested):
Very good for the PR message! 👍 
Small comments and nits for the code but you're on tracks!

OmarKhalil2001 committed:
[IMP] Added contraints to Estate Module. Constraints dictate that prices must be positive. names, tags, and building types must be unique. Offer must be at least 90% of buildings value.

OmarKhalil2001 committed:
[CLN] Module: Estate. Renamed files and attributes to follow the conventions.

barracudapps commented on estate/models/offer.py:
```diff
@@ -0,0 +1,85 @@
+from odoo import models, fields, api
+from odoo.exceptions import UserError
+from datetime import timedelta
+from odoo.tools.float_utils import float_compare
```
Nit here
```suggestion
from datetime import timedelta

from odoo import models, fields, api
from odoo.exceptions import UserError
from odoo.tools.float_utils import float_compare
```

barracudapps commented on estate/models/offer.py:
```diff
@@ -0,0 +1,85 @@
+
+
+class Offer(models.Model):
+    _name = "estate.offers"
```
Same as for buildings, make it consistent
```suggestion
    _name = "estate.offer"
```

barracudapps commented on estate/models/offer.py:
```diff
@@ -0,0 +1,85 @@
+            if record.status != "accepted" and record.building_id.state not in [
+                "sold",
+                "canceled",
+            ]:
```
IMHO this is more readable
```suggestion
            if record.status != 'accepted' and record.building_id.state not in ['sold', 'canceled']:
```

barracudapps commented on estate/models/offer.py:
```diff
@@ -0,0 +1,85 @@
+            elif record.building_id.state in ["sold", "canceled"]:
+                raise UserError(
+                    self.env._("Cannot accept offers for sold or canceled buildings.")
+                )
```
```suggestion
                raise UserError(self.env._("Cannot accept offers for sold or canceled buildings."))
```

barracudapps commented on estate/models/offer.py:
```diff
@@ -0,0 +1,85 @@
+                    0.9 * record.building_id.value, record.price, precision_digits=2
+                )
+                == 1
+            ):
```
As we know `float_compare` returns either `1`, `0` or `-1` you can save some characters to keep the condition on a single line
```suggestion
            if float_compare(0.9 * record.building_id.value, record.price, precision_digits=2) > 0:
```

barracudapps reviewed (changes requested):
Still some nits

OmarKhalil2001 committed:
[IMP] Module: Estate.
Description:
1. Add styling options to tags, form views.
2. Hide unnecessary fields where appropriate.
3. Add button to view offers for each property type.
4. Add color coding to property lists

OmarKhalil2001 committed:
[IMP] Module: Estate
    Description:
    1. Extended res.user with building_ids attribute
    2. reflected the new attribute in res.user form view

OmarKhalil2001 committed:
[ADD] Added Module: Estate Account
Description:
1. Added a new module to create invoices for new completed sales in estate module

OmarKhalil2001 committed:
[IMP] Module: Estate, Estate_accounting

Description:
1. Added Kanban view to estate module building model.
2. Group by building type.
3. formated estate_accounting's model estate_accounting

OmarKhalil2001 committed:
[CLN] Module: Estate, Estate_accounting

Description:
1. cleaning code to follow odoo's convention
2. addressing nit left by the PR reviewer
